### PR TITLE
[ROCm][FEAT] Integrate aiter gemm w8a8 ptpc

### PR DESCRIPTION
**Pull request by vllmellm**
## Purpose

This PR integrates aiter PTPC kernels to optimize FP8 inference performance on AMD ROCm platforms. By leveraging bpreshuffle and ck operators from the aiter library, we provide a significant performance uplift for mainstream Large Language Models compared to the standard torch.scaled_mm baseline.

Key enhancements:

Integrated bpreshuffle (Priority 1): Optimized weight-shuffling kernel for maximum throughput.

Integrated ck (Priority 2): High-performance Composable Kernel fallback when shuffling is not used.

Broad Model Support: Verified and tuned for Qwen3-8B, and Qwen2.5-Coder-32B.

Performance vs. Accuracy Trade-off: Achieved substantial throughput gains while keeping accuracy degradation (LMeval) within a reasonable and acceptable margin.

## Test Plan

1. Accuracy Evaluation (LMeval)
Commands used for gsm8k tasks:

Enable bpreshuffle: System defaults to bpreshuffle when `VLLM_ROCM_USE_AITER=1` is set and shapes are aligned.

Switch to ck: (Optional) Manually disabling shuffle logic or using untuned shapes defaults to CK path. core set like 
```
export VLLM_ROCM_USE_AITER=1
export VLLM_DISABLED_KERNELS="AiterPreshuffledPerTokenFp8ScaledMMLinearKernel"
```

2. Performance Benchmark (vllm bench serve)
Command example for DeepSeek-R1:

```
vllm bench serve \
    --backend vllm \
    --model "EmbeddedLLM/deepseek-r1-FP8-Dynamic" \
    --dataset-name random \
    --random-input-len 1000 \
    --random-output-len 1000 \
    --num-prompts 640 \
    --max-concurrency 64 \
    --save-result \
    --trust-remote-code \
    --port 8001 \
    --endpoint /v1/completions 
```

## Test Result

1. LMeval Accuracy Comparison
The following table summarizes the accuracy across different kernels. Minor score fluctuations observed with bpreshuffle are within the expected range for FP8 quantization.

Screenshot 2026-03-16 at 23 07 05

2. Benchmark Throughput Performance
Integrated aiter kernels show a significant tokens/sec improvement over the native Torch implementation. NOTE: These are the performance after tuning the kernels. Tuning procedure can be found in https://github.com/ROCm/aiter/tree/main/csrc/ck_gemm_a8w8_blockscale_bpreshuffle and https://github.com/ROCm/aiter/tree/main/csrc/ck_gemm_a8w8_blockscale

Screenshot 2026-03-16 at 23 07 48

### CODE REVIEW

File: vllm/_aiter_ops.py [hunk 1]
```diff
@@ -420,6 +420,34 @@ def _rocm_aiter_gemm_a8w8_fake(
...
+    As: torch.Tensor | None = None,
+    Bs: torch.Tensor | None = None,
+    bias: torch.Tensor | None = None,
+    output_dtype: torch.dtype = torch.float16,
+) -> torch.Tensor:
+    from aiter import gemm_a8w8_bpreshuffle
+    output = gemm_a8w8_bpreshuffle(A, B, As, Bs, None, output_dtype)
+    if bias is not None:
+        output.add_(bias)
+    return output
```
[the remark this answers is not in the document]
Comment (vllmellm) [reply vllm/_aiter_ops.py:539]: Regarding the bias fusion: Current aiter implementation of gemm_a8w8_bpreshuffle has an explicit assertion assert bias is None. We cannot fuse the bias operation until aiter supports it.

File: vllm/_aiter_ops.py [hunk 2]
```diff
@@ -420,6 +420,34 @@ def _rocm_aiter_gemm_a8w8_fake(
     return Y
 
 
+def _rocm_aiter_gemm_a8w8_bpreshuffle_impl(
+    A: torch.Tensor,
+    B: torch.Tensor,
+    As: torch.Tensor | None = None,
+    Bs: torch.Tensor | None = None,
```
[the remark this answers is not in the document]
Comment (vllmellm) [reply vllm/_aiter_ops.py]: done

File: tests/utils.py [hunk 3]
```diff
@@ -1373,9 +1373,10 @@ def __init__(
         force_kernel: FP8ScaledMMLinearKernel | None = None,
     ):
         super().__init__()
+        N, K = weight_shape
         per_tensor_weights = weight_quant_key.scale.group_shape.is_per_tensor()
         is_static_activation_scale = activation_quant_key.scale.static
-        weight_scale_shape = (1,) if per_tensor_weights else (weight_shape[0], 1)
+        weight_scale_shape = (1,) if per_tensor_weights else (N, 1)
```
Comment (tjtanaa) [inline tests/utils.py]: Is it necessary to change this?

File: vllm/_aiter_ops.py [hunk 4]
```diff
@@ -1087,6 +1116,14 @@ def register_ops_once() -> None:
                 dispatch_key=current_platform.dispatch_key,
             )
 
+            direct_register_custom_op(
+                op_name="rocm_aiter_gemm_a8w8_bpreshuffle",
+                op_func=_rocm_aiter_gemm_a8w8_bpreshuffle_impl,
+                mutates_args=[],
```
Comment (tjtanaa) [inline vllm/_aiter_ops.py]: we don't need to define `mutates_args=[],` if it is empty

File: vllm/_aiter_ops.py [hunk 5]
```diff
@@ -1087,6 +1116,14 @@ def register_ops_once() -> None:
...
+                fake_impl=_rocm_aiter_gemm_a8w8_bpreshuffle_fake,
+                dispatch_key=current_platform.dispatch_key,
```
Comment (tjtanaa) [inline vllm/_aiter_ops.py]: I think we don't need to define this `dispatch_key` anymore.

File: vllm/_aiter_ops.py (hunk 4 above)
Comment (vllmellm) [reply vllm/_aiter_ops.py]: Hi @tjtanaa , Thanks for your quickly review, I will update it one by one.

File: tests/utils.py (hunk 3 above)
Comment (vllmellm) [reply tests/utils.py]: not necessary, It has been modified.

File: vllm/_aiter_ops.py (hunk 5 above)
Comment (vllmellm) [reply vllm/_aiter_ops.py]: done

File: vllm/_aiter_ops.py [hunk 6]
```diff
@@ -456,6 +480,35 @@ def _rocm_aiter_gemm_a8w8_fake(
     return Y
 
 
+def _rocm_aiter_shuffled_per_token_w8a8_gemm_impl(
```
Comment (tjtanaa) [inline vllm/_aiter_ops.py]: Let's keep the semantic, use `preshuffled` rather than just `shuffled`

Comment (vllmellm) [reply vllm/_aiter_ops.py]: yes, done. Thanks

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 7]
```diff
@@ -194,12 +195,13 @@ def _add_fused_moe_to_target_scheme_map(self):
...
-        if (
-            "Linear" not in self.target_scheme_map
-            or "FusedMoE" in self.target_scheme_map
-        ):
+        if "Linear" not in self.target_scheme_map:
             return
-        self.target_scheme_map["FusedMoE"] = self.target_scheme_map["Linear"]
+        if "FusedMoE" not in self.target_scheme_map:
+            self.target_scheme_map["FusedMoE"] = self.target_scheme_map["Linear"]
+        for ds_target in ["DeepSeekV2FusedQkvAProj", "DeepSeekV3FusedQkvAProj"]:
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py]: Please check if this is still needed? I think this has been resolved by this PR https://github.com/vllm-project/vllm/pull/36247

Comment (vllmellm) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py]: yes, I have checked it, this logic ensure The deepseek-r1 model was successfully load and start. Previously, #36247 resolved the issue of abnormal values ​​during deepseek-r1 model inference. we should Keep this logic.

**Review by tjtanaa — APPROVED** [review]
LGTM

File: vllm/model_executor/layers/quantization/__init__.py [hunk 8]
```diff
@@ -12,6 +12,8 @@
 QuantizationMethods = Literal[
     "awq",
     "fp8",
+    "mxfp8",
+    "ptpc_fp8",
```
Comment (BadrBasowid) [inline vllm/model_executor/layers/quantization/__init__.py]: why is this needed?

File: vllm/model_executor/layers/quantization/__init__.py [hunk 9]
```diff
@@ -128,13 +129,14 @@ def get_quantization_config(quantization: str) -> type[QuantizationConfig]:
...
     from .moe_wna16 import MoeWNA16Config
     from .mxfp4 import Mxfp4Config
-    from .mxfp8 import Mxfp8Config
     from .petit import PetitNvFp4Config
     from .torchao import TorchAOConfig
 
     method_to_config: dict[str, type[QuantizationConfig]] = {
         "awq": AWQConfig,
         "fp8": Fp8Config,
+        "ptpc_fp8": Fp8Config,
```
Comment (BadrBasowid) [inline vllm/model_executor/layers/quantization/__init__.py]: also here, why is this needed?

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 10]
```diff
@@ -1148,13 +1154,12 @@ def create_weights(
         layer.register_parameter("weight_scale", weight_scale)
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
-        # Rename ModelOpt checkpoint names to standardized names
```
Comment (BadrBasowid) [inline vllm/model_executor/layers/quantization/modelopt.py:1179]: what is the purpose of this change?

File: vllm/model_executor/layers/quantization/__init__.py (hunk 9 above)
Comment (vllmellm) [reply vllm/model_executor/layers/quantization/__init__.py]: Because the upstream ptpc_fp8 file has been removed, it needs to be added when resolving conflicts.

File: vllm/model_executor/layers/quantization/modelopt.py (hunk 10 above)
Comment (vllmellm) [reply vllm/model_executor/layers/quantization/modelopt.py:1179]: This was caused by frequent merging of the main branch and has been restored.

File: vllm/model_executor/layers/quantization/__init__.py (hunk 9 above)
Comment (BadrBasowid) [reply vllm/model_executor/layers/quantization/__init__.py]: I think it should be reverted. The `ptpc_fp8` method was only needed for dynamic quantization, if i am not mistaken, and the class itself has been removed.

File: vllm/model_executor/layers/quantization/__init__.py (hunk 8 above)
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/__init__.py]: @vllmellm please remove this. `ptpc_fp8` has been removed.

File: vllm/model_executor/layers/quantization/__init__.py [hunk 11]
```diff
@@ -128,13 +129,14 @@ def get_quantization_config(quantization: str) -> type[QuantizationConfig]:
...
+        "mxfp8": ModelOptMxFp8Config,
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/__init__.py]: remove this changes `"mxfp8": Mxfp8Config,`

File: vllm/model_executor/layers/quantization/__init__.py (hunk 8 above)
Comment (vllmellm) [reply vllm/model_executor/layers/quantization/__init__.py]: done, removed.

File: vllm/model_executor/layers/quantization/__init__.py (hunk 11 above)
Comment (vllmellm) [reply vllm/model_executor/layers/quantization/__init__.py]: done, removed.

**Review by tjtanaa — APPROVED** [review]
LGTM. Thanks for keeping the PR updated.
